data/abi/abi_encode_test.go

```diff
@@ -0,0 +1,1016 @@
+		require.NoError(t, err, "make static bool array should not return error")
+		boolArrDynamicVal, err := MakeDynamicArray(boolValArr, MakeBoolType())
+		require.NoError(t, err, "make dynamic bool array should not return error")
+		tupleVal, err := MakeTuple([]Value{boolArrStaticVal, boolArrDynamicVal})
+		require.NoError(t, err, "make tuple for static/dynamic bool array should not return error")
+		encodedInput := []byte{
+			0b11000000,
+			0x00, 0x03,
+			0x00, 0x02, 0b11000000,
+		}
+		decoded, err := Decode(encodedInput, Type{
+			enumIndex:    Tuple,
+			staticLength: 2,
+			childTypes: []Type{
+				{
+					enumIndex:    ArrayStatic,
+					staticLength: 2,
+					childTypes:   []Type{MakeBoolType()},
+				},
+				{
+					enumIndex:  ArrayDynamic,
+					childTypes: []Type{MakeBoolType()},
+				},
+			},
+		})
+		require.NoError(t, err, "decode tuple for static/dynamic bool array should not return error")
+		require.Equal(t, tupleVal, decoded, "decoded tuple value do not match with expected")
+	})
+
+	// decoding test for tuple with all dynamic bool array
+	// expected value: (bool[], bool[])
+	// expected value: ({}, {})
+	/*
+	   input bytes:
+	   0x00, 0x04      (first dynamic bool array starts at 4th byte)
+	   0x00, 0x06      (second dynamic bool array starts at 6th byte)
+	   0x00, 0x00      (first dynamic bool array length 0)
+	   0x00, 0x00      (second dynamic bool array length 0)
+	*/
+	t.Run("empty dynamic array tuple decoding", func(t *testing.T) {
+		emptyDynamicArray, err := MakeDynamicArray([]Value{}, MakeBoolType())
+		require.NoError(t, err, "make empty dynamic array should not return error")
+		tupleVal, err := MakeTuple([]Value{emptyDynamicArray, emptyDynamicArray})
+		require.NoError(t, err, "make empty dynamic array tuple should not return error")
+		encodedInput := []byte{
+			0x00, 0x04, 0x00, 0x06,
+			0x00, 0x00, 0x00, 0x00,
+		}
+		decoded, err := Decode(encodedInput, Type{
+			enumIndex:    Tuple,
+			staticLength: 2,
+			childTypes: []Type{
+				{
+					enumIndex:  ArrayDynamic,
+					childTypes: []Type{MakeBoolType()},
+				},
+				{
+					enumIndex:  ArrayDynamic,
+					childTypes: []Type{MakeBoolType()},
+				},
+			},
+		})
+		require.NoError(t, err, "decode tuple for empty dynamic array should not return error")
+		require.Equal(t, tupleVal, decoded, "decoded tuple value do not match with expected")
+	})
+
+	// decoding test for empty tuple
+	// expected value: ()
+	// byte input: ""
+	t.Run("empty tuple decoding", func(t *testing.T) {
+		emptyTuple, err := MakeTuple([]Value{})
+		require.NoError(t, err, "make empty tuple should not return error")
+		encodedInput := make([]byte, 0)
+		decoded, err := Decode(encodedInput, Type{
+			enumIndex:    Tuple,
+			staticLength: 0,
+			childTypes:   []Type{},
+		})
+		require.NoError(t, err, "decode empty tuple should not return error")
+		require.Equal(t, emptyTuple, decoded, "empty tuple encode should not return error")
+	})
+}
+
+func TestDecodeInvalid(t *testing.T) {
```

Comment:
Hand written some invalid test cases for decoding, by corrupting the bytes of the ABI value.

The goal is to see decoding function is robust enough to accept erroneous bytes (without crashing itself) and distinguish if the byte match with the ideal ABI type.

Assuming that all the ABI value to be encoded are well-formed, I did not provide the test `TestEncodeValid`, but based on the current implementation of `Encode`, ill-formed ABI values will be eliminated by `Get` method in ABI value.